File: Java/Leetcode/Greedy/JumpGame2.java
package com.brianprojects.Leetcode.Greedy;

public class JumpGame2 {
    //Time Complexity: O(n) || Space Complexity: O(1)
    public int jump(int[] nums){
        int jumps = 0, curEnd = 0, curFarthest = 0;
        for (int i = 0; i < nums.length - 1; i++) {
            curFarthest = Math.max(curFarthest, i + nums[i]);
            if (i == curEnd) {
                jumps++;
                curEnd = curFarthest;
            }
        }
        return jumps;
    }
}

File: Python/Leetcode_Python/Greedy/JumpGame2.py
# Time Complexity: O(n) || Space Complexity: O(n)
from typing import List


def jump(nums: List[int]) -> int:
    if not nums:
        return 0

    jumps = 0
    end = 0
    farth = 0

    for i in range(len(nums) - 1):
        farth = max(farth, i + nums[i])
        print(f"Farth: {farth}, Jump: {jumps}, end:{end}, i:{i}")
        if i == end:
            jumps += 1
            end = farth
    return jumps


jump([1, 2, 1, 1, 1])
